File: dikta-macos/Dikta/Models/Language.swift
import Foundation

/// Supported languages for dictation
enum Language: String, Codable, CaseIterable {
    case english = "en"
    case swedish = "sv"
    case indonesian = "id"
    case spanish = "es"
    case french = "fr"
    case german = "de"
    case portuguese = "pt"
    case italian = "it"
    case dutch = "nl"
    case finnish = "fi"
    case norwegian = "no"
    case danish = "da"

    var displayName: String {
        switch self {
        case .english: return "English"
        case .swedish: return "Svenska"
        case .indonesian: return "Bahasa Indonesia"
        case .spanish: return "Español"
        case .french: return "Français"
        case .german: return "Deutsch"
        case .portuguese: return "Português"
        case .italian: return "Italiano"
        case .dutch: return "Nederlands"
        case .finnish: return "Suomi"
        case .norwegian: return "Norsk"
        case .danish: return "Dansk"
        }
    }

    /// Short code for menu bar display
    var menuBarCode: String {
        rawValue.uppercased()
    }

    /// Whisper language code
    var whisperCode: String {
        rawValue
    }

    /// Next language in the cycle
    var next: Language {
        let all = Language.allCases
        let index = all.firstIndex(of: self)!
        return all[(index + 1) % all.count]
    }
}

File: dikta-windows/Models/Language.cs
namespace DiktaWindows.Models;

/// <summary>Supported languages for dictation — mirrors the macOS Language enum.</summary>
public sealed class Language
{
    public static readonly Language English    = new("en", "English",          "EN");
    public static readonly Language Swedish    = new("sv", "Svenska",          "SV");
    public static readonly Language Indonesian = new("id", "Bahasa Indonesia", "ID");
    public static readonly Language Spanish    = new("es", "Español",          "ES");
    public static readonly Language French     = new("fr", "Français",         "FR");
    public static readonly Language German     = new("de", "Deutsch",          "DE");
    public static readonly Language Portuguese = new("pt", "Português",        "PT");
    public static readonly Language Italian    = new("it", "Italiano",         "IT");
    public static readonly Language Dutch      = new("nl", "Nederlands",       "NL");
    public static readonly Language Finnish    = new("fi", "Suomi",            "FI");
    public static readonly Language Norwegian  = new("no", "Norsk",            "NO");
    public static readonly Language Danish     = new("da", "Dansk",            "DA");

    public static readonly IReadOnlyList<Language> All = new[]
    {
        English, Swedish, Indonesian, Spanish, French, German,
        Portuguese, Italian, Dutch, Finnish, Norwegian, Danish
    };

    /// <summary>Whisper language code passed to Whisper.net (e.g. "en", "sv").</summary>
    public string WhisperCode { get; }

    /// <summary>Native display name shown in the UI (e.g. "Svenska", "Français").</summary>
    public string DisplayName { get; }

    /// <summary>Short uppercase code for tray / menu bar display (e.g. "EN", "SV").</summary>
    public string MenuBarCode { get; }

    private Language(string whisperCode, string displayName, string menuBarCode)
    {
        WhisperCode = whisperCode;
        DisplayName = displayName;
        MenuBarCode = menuBarCode;
    }

    /// <summary>
    /// Returns the Language matching <paramref name="code"/>, or <see cref="English"/> if unknown.
    /// </summary>
    public static Language FromCode(string code) =>
        All.FirstOrDefault(l => l.WhisperCode == code) ?? English;
}
